// @mysten/walrus/dist/esm/contracts/wal/wal.js
import { MoveStruct, normalizeMoveArguments } from "../utils/index.js";
import { bcs } from "@mysten/sui/bcs";
import * as object from "./deps/sui/object.js";
const $moduleName = "@local-pkg/wal::wal";
const WAL = new MoveStruct({
  name: `${$moduleName}::WAL`,
  fields: {
    dummy_field: bcs.bool()
  }
});
const ProtectedTreasury = new MoveStruct({
  name: `${$moduleName}::ProtectedTreasury`,
  fields: {
    id: object.UID
  }
});
const TreasuryCapKey = new MoveStruct({
  name: `${$moduleName}::TreasuryCapKey`,
  fields: {
    dummy_field: bcs.bool()
  }
});
function totalSupply(options) {
  const packageAddress = options.package ?? "@local-pkg/wal";
  const argumentsTypes = [`${packageAddress}::wal::ProtectedTreasury`];
  const parameterNames = ["treasury"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "wal",
    function: "total_supply",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
function burn(options) {
  const packageAddress = options.package ?? "@local-pkg/wal";
  const argumentsTypes = [
    `${packageAddress}::wal::ProtectedTreasury`,
    `0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<${packageAddress}::wal::WAL>`
  ];
  const parameterNames = ["treasury", "coin"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "wal",
    function: "burn",
    arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
  });
}
export {
  ProtectedTreasury,
  TreasuryCapKey,
  WAL,
  burn,
  totalSupply
};
