// @mysten/seal/dist/esm/kdf.js
import { fromHex } from "@mysten/bcs";
import { sha3_256 } from "@noble/hashes/sha3";
import { G1Element } from "./bls12381.js";
import {
  ENCRYPTED_SHARE_LENGTH,
  flatten,
  KEY_LENGTH,
  MAX_U8,
  SUI_ADDRESS_LENGTH
} from "./utils.js";
const DST = new TextEncoder().encode("SUI-SEAL-IBE-BLS12381-00");
const KDF_DST = new TextEncoder().encode("SUI-SEAL-IBE-BLS12381-H2-00");
const DERIVE_KEY_DST = new TextEncoder().encode("SUI-SEAL-IBE-BLS12381-H3-00");
function hashToG1(id) {
  return G1Element.hashToCurve(flatten([DST, id]));
}
function kdf(element, nonce, id, objectId, index) {
  if (!Number.isInteger(index) || index < 0 || index > MAX_U8) {
    throw new Error(`Invalid index ${index}`);
  }
  const objectIdBytes = fromHex(objectId);
  if (objectIdBytes.length !== SUI_ADDRESS_LENGTH) {
    throw new Error(`Invalid object id ${objectId}`);
  }
  const hash = sha3_256.create();
  hash.update(KDF_DST);
  hash.update(element.toBytes());
  hash.update(nonce.toBytes());
  hash.update(hashToG1(id).toBytes());
  hash.update(objectIdBytes);
  hash.update(new Uint8Array([index]));
  return hash.digest();
}
var KeyPurpose = /* @__PURE__ */ ((KeyPurpose2) => {
  KeyPurpose2[KeyPurpose2["EncryptedRandomness"] = 0] = "EncryptedRandomness";
  KeyPurpose2[KeyPurpose2["DEM"] = 1] = "DEM";
  return KeyPurpose2;
})(KeyPurpose || {});
function tag(purpose) {
  switch (purpose) {
    case 0 /* EncryptedRandomness */:
      return new Uint8Array([0]);
    case 1 /* DEM */:
      return new Uint8Array([1]);
    default:
      throw new Error(`Invalid key purpose ${purpose}`);
  }
}
function deriveKey(purpose, baseKey, encryptedShares, threshold, keyServers) {
  if (!Number.isInteger(threshold) || threshold <= 0 || threshold > MAX_U8) {
    throw new Error(`Invalid threshold ${threshold}`);
  }
  if (encryptedShares.length !== keyServers.length) {
    throw new Error(
      `Mismatched shares ${encryptedShares.length} and key servers ${keyServers.length}`
    );
  }
  const keyServerBytes = keyServers.map((keyServer) => fromHex(keyServer));
  if (keyServerBytes.some((keyServer) => keyServer.length !== SUI_ADDRESS_LENGTH)) {
    throw new Error(`Invalid key servers ${keyServers}`);
  }
  if (encryptedShares.some((share) => share.length !== ENCRYPTED_SHARE_LENGTH)) {
    throw new Error(`Invalid encrypted shares ${encryptedShares}`);
  }
  if (baseKey.length !== KEY_LENGTH) {
    throw new Error(`Invalid base key ${baseKey}`);
  }
  const hash = sha3_256.create();
  hash.update(DERIVE_KEY_DST);
  hash.update(baseKey);
  hash.update(tag(purpose));
  hash.update(new Uint8Array([threshold]));
  encryptedShares.forEach((share) => hash.update(share));
  keyServerBytes.forEach((keyServer) => hash.update(keyServer));
  return hash.digest();
}
export {
  KeyPurpose,
  deriveKey,
  hashToG1,
  kdf
};
